### PR TITLE
feat: 近隣検索 API を実装 (#117)

### LINKED ISSUE #117: [Rails] 近隣検索 API の実装
**ラベル**: `rails`, `api`
**依存**: #113
**対応するフロント側 issue**: matcha-to-jinja#17

## 概要

現在地から半径 N km 以内の抹茶店・神社をまとめて返す。

## エンドポイント

```
GET /api/v1/nearby?lat=35.003&lng=135.771&radius=1.5
```

- `radius` 単位は **km**（フロント `NearbyMap` と合わせる）
- `lat` / `lng` が欠損 / 非数なら **400** を返す（フロント mock #32 で同様の検証を入れている）
- `radius` 省略時は 1.5km

## レスポンス

```json
{
  "greenteas": [
    { "id": 1, "name": "...", "latitude": 35.003, "longitude": 135.771, "distance_meters": 450 }
  ],
  "temples": [
    { "id": 3, "name": "...", "latitude": 35.005, "longitude": 135.770, "distance_meters": 800 }
  ]
}
```

- 各配列は `distance_meters` 昇順
- スポット数は各 50 件で打ち切り（パフォーマンス保護）

## タスク

- [ ] `Api::V1::NearbyController#search`
- [ ] Geokit による距離計算（必要なら `acts_as_mappable` を Greentea/Temple に追加）
- [ ] パラメータ検証（`lat` / `lng` の有限数判定、`radius` の上限 = 例: 10km）
- [ ] 専用 serializer（距離付きの軽量版）
- [ ] N+1 を避ける（`includes` 等）

## 受け入れ条件

- フロント `/nearby`（matcha-to-jinja PR #40）が実 API でそのまま動作する
- `lat=abc` で 400 が返る

## 参考

- matcha-to-jinja: `docs/migration-plan.md` 1-3 (`/nearby`)

### PR DESCRIPTION
**Pull request by hiiragi17**
## 概要

Closes #117

現在地から半径 N km 以内の抹茶店・神社をまとめて返す `GET /api/v1/nearby` を実装した。
フロント matcha-to-jinja の `NearbyMap` （matcha-to-jinja#17）から実 API でそのまま叩ける契約。

```
GET /api/v1/nearby?lat=34.9676&lng=135.7741&radius=1.5
```

### レスポンス例

```json
{
  "greenteas": [
    { "id": 1, "name": "...", "latitude": 34.9676, "longitude": 135.7741, "distance_meters": 0 }
  ],
  "temples": [
    { "id": 3, "name": "...", "latitude": 34.9678, "longitude": 135.7742, "distance_meters": 28 }
  ]
}
```

## 実装方針

- `Api::V1::NearbyController#search` を追加（未認証アクセス可）
- `Greentea` / `Temple` ともに `acts_as_mappable` が入っているので `within(radius, origin:).by_distance(origin:)` で距離付き・距離昇順のスコープを構築し、各 50 件で打ち切り
- 距離計算は既存の `NearbyGreenteaSerializer` / `NearbyTempleSerializer` をそのまま再利用（`obj.distance` が `within` から伝搬する）
- `lat` / `lng` のパースは `Float()` の例外ベース。`nil` / 数値以外 / 非有限 / 範囲外（`|lat| > 90` / `|lng| > 180`）は 400
- `radius` 省略時は **1.5km**、上限 **10km**。0 以下や上限超過は 400
- 結果なしは `{"greenteas": [], "temples": []}`

## 確認方法

```bash
bundle exec rspec spec/requests/api/v1/nearby_spec.rb
```

ローカルで疎通確認:

```bash
bin/rails server -p 3001
curl 'http://localhost:3001/api/v1/nearby?lat=34.9676&lng=135.7741'
curl 'http://localhost:3001/api/v1/nearby?lat=34.9676&lng=135.7741&radius=5'
curl -i 'http://localhost:3001/api/v1/nearby?lat=abc&lng=135.7741'   # → 400
```

## 影響範囲

- 新規エンドポイント追加のみ。既存 Web 画面 / 他の API には影響なし
- `Api::V1::BaseController` の `flatten_serialized` / `render_bad_request` を流用しており、共通基盤に変更はない

## チェックリスト

- [x] request spec を追加（正常系 / 異常系 / ソート / 半径 / 上限 50 件 / 空配列）
- [x] フロントの `docs/migration-plan.md` 1-3 のレスポンス契約と一致
- [x] 既存 `NearbyGreenteaSerializer` / `NearbyTempleSerializer` を再利用

## コメント

- 50 件打ち切りは `MAX_RESULTS` 定数で保護しており、spec では `stub_const` で 2 件に絞って検証
- `radius` 上限 10km はパフォーマンス保護の目安。フロント要件が変われば後続で調整可能

---
_Generated by [Claude Code](https://claude.ai/code/session_01VCgwcj6oMkbCgnWYvX1sbP)_

## Release Notes

* **New Features**
  * Added a nearby location search API endpoint accepting latitude, longitude, and optional radius parameters, returning nearby places sorted by distance with automatic result limiting and distance calculations.

* **Tests**
  * Added comprehensive request tests covering valid inputs, validation edge cases, and distance computation.

### CODE REVIEW

File: spec/requests/api/v1/nearby_spec.rb [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      end
+
+      it 'computes distance_meters consistent with Geokit' do
+        get '/api/v1/nearby', params: { lat: origin_lat, lng: origin_lng }
+
+        origin = Geokit::LatLng.new(origin_lat, origin_lng)
+        expected = (origin.distance_to(Geokit::LatLng.new(near_greentea.latitude, near_greentea.longitude)) * 1000).round
+        actual = response.parsed_body['greenteas'].find { |g| g['id'] == near_greentea.id }['distance_meters']
+        expect(actual).to be_within(5).of(expected)
+      end
```
[the remark this answers is not in the document]
Comment (hiiragi17) [reply spec/requests/api/v1/nearby_spec.rb:83]: a3409ff で対応しました。`units: :kms` を明示するよう修正しています。

ご指摘どおり `geokit-rails` はモデル側でしか default_units を設定しておらず（`acts_as_mappable default_units: :kms`）、`Geokit::LatLng#distance_to` を直接呼ぶケースではグローバル設定に依存してしまうため、明示しておく方が堅牢です。

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by hiiragi17** [comment]
@codex review
